In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
import sklearn.metrics as sm
from sklearn import tree
from sklearn import preprocessing

df = pd.read_csv('wdbc.csv')

In [2]:
# check data type
df.dtypes

842302        int64
M            object
17.99       float64
10.38       float64
122.8       float64
1001        float64
0.1184      float64
0.2776      float64
0.3001      float64
0.1471      float64
0.2419      float64
0.07871     float64
1.095       float64
0.9053      float64
8.589       float64
153.4       float64
0.006399    float64
0.04904     float64
0.05373     float64
0.01587     float64
0.03003     float64
0.006193    float64
25.38       float64
17.33       float64
184.6       float64
2019        float64
0.1622      float64
0.6656      float64
0.7119      float64
0.2654      float64
0.4601      float64
0.1189      float64
dtype: object

In [3]:
# check NaN
df.isnull().any()

842302      False
M           False
17.99       False
10.38       False
122.8       False
1001        False
0.1184      False
0.2776      False
0.3001      False
0.1471      False
0.2419      False
0.07871     False
1.095       False
0.9053      False
8.589       False
153.4       False
0.006399    False
0.04904     False
0.05373     False
0.01587     False
0.03003     False
0.006193    False
25.38       False
17.33       False
184.6       False
2019        False
0.1622      False
0.6656      False
0.7119      False
0.2654      False
0.4601      False
0.1189      False
dtype: bool

In [4]:
# B=>0 M=>1
def transform(x):
    for i in range(0,len(x)):
        if(x[i]=='B'):
            x[i]=0;
        else:
            x[i]=1;

In [8]:
x = df.iloc[:,2:31].values
y = df.iloc[:,1].values
score_train=[]
score_test=[]
# Bagging
for i in range(0,20):
    (x_train,x_test,y_train,y_test)=train_test_split(x, y, test_size=0.3)
    clf = DecisionTreeClassifier()
    clf.fit(x_train, y_train)
    clf.score(x_test, y_test)
    y_pred_DT1=clf.predict(x_test)
    score_test.append(round(clf.score(x_test, y_test),5))
    score_train.append(round(clf.score(x_train, y_train),5))
accurancy_test=np.average(score_test)
accurancy_train=np.average(score_train)    
print("For DT1:\nTrain accurancy=%f\nTest accurancy=%f"%(accurancy_train,accurancy_test))
transform(y_test)
transform(y_pred_DT1)
precision_DT1=sm.precision_score(y_test.astype(int),y_pred_DT1.astype(int))
recall_DT1=sm.recall_score(y_test.astype(int),y_pred_DT1.astype(int))
print("Precision=%f"%precision_DT1)
print("Recall=%f"%recall_DT1)

For DT1:
Train accurancy=1.000000
Test accurancy=0.929531
Precision=0.965517
Recall=0.888889


In [14]:
# Bagging
score_train=[]
score_test=[]
for turns in range(0,20):
    (x_train,x_test,y_train,y_test)=train_test_split(x, y, test_size=0.3)
    clf_with_limit = DecisionTreeClassifier(max_depth=8)
    clf_with_limit.fit(x_train, y_train)
    y_pred_DT2=clf.predict(x_test)
    score_test.append(round(clf.score(x_test, y_test),5))
    score_train.append(round(clf.score(x_train, y_train),5))
accurancy_test=np.average(score_test)
accurancy_train=np.average(score_train)
print("For DT2:\nTrain accurancy=%f\nTest accurancy=%f"%(accurancy_train,accurancy_test))
transform(y_test)
transform(y_pred_DT2)
precision_DT2=sm.precision_score(y_test.astype(int),y_pred_DT2.astype(int))
recall_DT2=sm.recall_score(y_test.astype(int),y_pred_DT2.astype(int))
print("Precision=%f"%precision_DT2)
print("Recall=%f"%recall_DT2)

For DT2:
Train accurancy=0.983882
Test accurancy=0.984796
Precision=0.984848
Recall=0.970149


__DT1 Train Accurancy > DT2 Train Accurancy__ <br>
__DT1 Test Accurancy < DT2 Test Accurancy__ <br>
__So in DT1 decision tree is over-fitting which is fit to noise. After limitting the maximum depth, the DT2 decision tree is more great.__


||Accuracy|Accuracy|Precision|Recall|
|---|---|---|---|---|
||Train|Test|||
|DT1|1.000|0.929|0.966|0.900|
|DT2|0.984|0.985|0.985|0.970|